plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java

```diff
@@ -346,102 +419,285 @@ private void reorderCapellaDeactivationMessages(Scenario scenario,
+    if (elementFromXtext instanceof org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage)
+      return ((org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage) elementFromXtext).getReturn() != null;
+    return false;
+  }
+  
+  protected static boolean hasExecution(Message elementFromXtext) {
+    if (elementFromXtext instanceof org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage) {
+      return ((org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage) elementFromXtext).getExecution() != null;
+    }
+    if (elementFromXtext instanceof org.polarsys.capella.scenario.editor.dsl.textualScenario.ArmTimerMessage) {
+      return ((org.polarsys.capella.scenario.editor.dsl.textualScenario.ArmTimerMessage) elementFromXtext).getExecution() != null;
+    }
+    return false;
+  }
+
+  private static void editStateFragment(Scenario scenario,
+      org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment xtextElement,
+      List<EObject> previousEditorStateFragments) {
+
+    InstanceRole instanceRole = EmbeddedEditorInstanceHelper.getInstanceRole(xtextElement.getTimeline());
+    EList<TimeLapse> ownedTimeLapses = scenario.getOwnedTimeLapses();
+
+    List<EObject> simillarStateFragments = previousEditorStateFragments.stream()
+        .filter(x -> ((org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) x).getTimeline()
+            .equals(xtextElement.getTimeline())
+            && ((org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) x).getName()
+                .equals(xtextElement.getName())
+            && ((org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) x).getKeyword()
+                .equals(xtextElement.getKeyword()))
+        .collect(Collectors.toList());
+
+    EObject relatedElement = getRelatedElement(scenario, instanceRole, xtextElement);
+
+    if (relatedElement == null)
+      return;
+
+    List<TimeLapse> capellaStateFragments = getStateFragmentsWithGivenInstanceRoleAndRelatedElement(scenario,
+        instanceRole, xtextElement, relatedElement);
+
+    if (capellaStateFragments.isEmpty() || capellaStateFragments.size() < simillarStateFragments.size()) {
+      InteractionState interactionStateStart = createInteractionState(DslConstants.START, relatedElement, instanceRole);
+      InteractionState interactionStateEnd = createInteractionState(DslConstants.FINISH, relatedElement, instanceRole);
+
+      scenario.getOwnedInteractionFragments().add(interactionStateStart);
+      scenario.getOwnedInteractionFragments().add(interactionStateEnd);
+
+      StateFragment stateFragment = createStateFragment(interactionStateStart, interactionStateEnd, relatedElement,
+          xtextElement.getName());
+
+      ownedTimeLapses.add(stateFragment);
+    }
+  }
+
+  private static List<TimeLapse> getStateFragmentsWithGivenInstanceRoleAndRelatedElement(Scenario scenario,
```

Comment:
Could you get a shorter name for the method and document what it does in the comment?